modules/http_push.c

```diff
@@ -61,30 +63,60 @@ static int nw_http_push_start_acquire_data(struct nodewatcher_module *module,
           /* Default. */
           timeout = 5;
         }
+
+        const char* data_string = json_object_to_json_string(data); // TODO: const?
+
         curl_easy_setopt(curl, CURLOPT_TIMEOUT, timeout);
         curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1);
         curl_easy_setopt(curl, CURLOPT_URL, url);
         curl_easy_setopt(curl, CURLOPT_POST, 1);
-        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_object_to_json_string(data));
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data_string);
+
 #if LIBCURL_VERSION_NUM >= 0x072700
         /* Pin server-side public key when configured. */
         char *server_pubkey = nw_uci_get_string(uci, "nodewatcher.@agent[0].push_server_pubkey");
+        char *auth_type = nw_uci_get_string(uci, "nodewatcher.@agent[0].authentication_method");
+        
         if (server_pubkey) {
-          curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0);
-          curl_easy_setopt(curl, CURLOPT_PINNEDPUBLICKEY, server_pubkey);
+          if (strcmp(auth_type, "hmac") == 0) {
+            unsigned char *hmac_result = HMAC(EVP_sha256(), server_pubkey, strlen(server_pubkey), (unsigned char *)data_string, strlen(data_string), NULL, NULL);
```

Comment:
I see, so you are using `server_pubkey` as the secret key for HMAC. This is not good, you should instead introduce a new configuration option for this new key.